### PR TITLE
Twisted backend

### PR DESCRIPTION
**Pull request by ghabbenjansen**
Implements #272

### CODE REVIEW

File: Tribler/Core/Modules/restapi/trustchain_endpoint.py [hunk 1]
```diff
@@ -318,14 +320,34 @@ def render_GET(self, request):
...
-
-        return json.dumps({"user_node": user_node,
-                           "focus_node": focus_node,
-                           "neighbor_level": neighbor_level,
-                           "nodes": nodes,
-                           "edges": edges})
+        # nodes, edges = tribler_chain_community.get_graph(focus_node, neighbor_level, max_neighbors,
+        #                                                  mandatory_nodes)
+        #
+        # return json.dumps({"user_node": user_node,
```
Comment (MaxVanDeursen) [inline Tribler/Core/Modules/restapi/trustchain_endpoint.py]: Remove commented code

File: Tribler/Core/Modules/restapi/trustchain_endpoint.py [hunk 2]
```diff
@@ -10,6 +10,8 @@
 from Tribler.community.triblerchain.community import TriblerChainCommunity
 from Tribler.Core.exceptions import OperationNotEnabledByConfigurationException
 
+NOT_DONE_YET = 1
```
Comment (MaxVanDeursen) [inline Tribler/Core/Modules/restapi/trustchain_endpoint.py]: Why is this constant set?

File: TriblerGUI/widgets/network_explorer/index.html [hunk 3]
```diff
@@ -10,7 +10,7 @@
 <body>
 
 <!-- GRAPH -->
-<svg id="graph" width="600" height="600" viewBox="0 0 600 600"></svg>
+<svg id="graph" width="600" height="600" viewBox="0 0 600 600" preserveAspectRatio="xMinYMid meet"></svg>
```
Comment (MaxVanDeursen) [inline TriblerGUI/widgets/network_explorer/index.html]: Please remove this GUI line

File: Tribler/Core/Modules/restapi/trustchain_endpoint.py [hunk 4]
```diff
@@ -318,14 +320,34 @@ def render_GET(self, request):
...
+            request.write(json.dumps({"user_node": user_node,
+                                      "focus_node": focus_node,
+                                      "neighbor_level": neighbor_level,
+                                      "nodes": nodes,
+                                      "edges": edges}))
+            request.finish()
+
+        d = tribler_chain_community.get_graph(focus_node, neighbor_level, max_neighbors, mandatory_nodes)
+        d.addCallback(finalize_request)
+        return NOT_DONE_YET
```
Comment (vandenheuvel) [inline Tribler/Core/Modules/restapi/trustchain_endpoint.py]: You could add an extra newline in here.

File: Tribler/community/triblerchain/database.py [hunk 5]
```diff
@@ -172,7 +174,12 @@ def get_graph_edges(self, public_key, neighbor_level=1):
             ORDER BY (ta.traffic_a_to_b + ta.traffic_b_to_a) DESC
         """ % {"query": query, "traffic": total_traffic}
 
-        return self.execute(query, (buffer(public_key), buffer(public_key))).fetchall()
+        def get_rows(result):
+            return result.fetchall()
+
+        d = task.deferLater(reactor, 0.0, self.execute, query, (buffer(public_key), buffer(public_key)))
```
Comment (vandenheuvel) [inline Tribler/community/triblerchain/database.py]: This looks a lot like a hack to me. What is happening here?

File: Tribler/community/triblerchain/community.py [hunk 6]
```diff
@@ -195,20 +193,19 @@ def get_node(self, public_key, nodes, total_up=None, total_down=None, total_neig
                 return {"total_up": total_traffic[0], "total_down": total_traffic[1],
                         "total_neighbors": total_traffic[2]}
 
-    def format_edges(self, public_key, neighbor_level):
+    def format_edges(self, result, public_key):
```
Comment (vandenheuvel) [inline Tribler/community/triblerchain/community.py]: Please rename this `result` variable.

File: Tribler/community/triblerchain/database.py (hunk 5 above)
Comment (ghabbenjansen) [inline Tribler/community/triblerchain/database.py]: This creates a task on the same thread as the thread this call is made on. The delay is set to 0.0, so the task is started instantly. The alternative would be to use `threads.deferToThreads`, but that won't work as the chance is high that the task will be scheduled on a different thread than the database.

Comment (vandenheuvel) [inline Tribler/community/triblerchain/database.py]: @MaxVanDeursen Do you know how this is supposed to work?

**Pull request merged**